Content.Server/Roles/RoleTimerSystem.cs

```diff
@@ -0,0 +1,306 @@
+                break;
+        }
+    }
+
+    private void OnRoundEnd(RoundRestartCleanupEvent ev)
+    {
+        _autoSaveAccumulator = 0f;
+        FullSave();
+    }
+
+    public override void Update(float frameTime)
+    {
+        if (_ticker.RunLevel != GameRunLevel.InRound || _autosaveDelay < 1f)
+        {
+            _autoSaveAccumulator = 0f;
+            return;
+        }
+
+        _autoSaveAccumulator += frameTime;
+
+        if (_autoSaveAccumulator < _autosaveDelay) return;
+
+        _autoSaveAccumulator -= _autosaveDelay;
+        FullSave();
+    }
+
+    private void OnUnAFK(ref UnAFKEvent ev)
+    {
+        _lastSetTime[ev.Session] = _timing.CurTime;
+    }
+
+    private void OnAFK(ref AFKEvent ev)
+    {
+        // Don't just write to the DB every time someone goes AFK.
+        Save(ev.Session, _timing.CurTime, false);
+    }
+
+    private void FullSave()
+    {
+        Sawmill.Info("Running full save of role timers");
+        var currentTime = _timing.CurTime;
+
+        // This is gonna have rounding if someone changes their jobs but it's only 5 minutes anyway, not like we need to track
+        // per second values every time they get a new job.
+        foreach (var player in Filter.GetAllPlayers())
+        {
+            var pSession = (IPlayerSession) player;
+            if (_afk.IsAfk(pSession)) continue;
+            Save(pSession, currentTime);
+            _roleTimers.SendRoleTimers(pSession);
+        }
+    }
+
+    public void Save(IPlayerSession pSession, TimeSpan currentTime, bool dbSave = true)
+    {
+        if (!_lastSetTime.TryGetValue(pSession, out var lastSave))
+        {
+            lastSave = currentTime;
+        }
+
+        if (currentTime <= lastSave) return;
+
+        var addedTime = currentTime - lastSave;
+        var roles = GetTimedRoles(pSession);
+        Sawmill.Info($"Adding {addedTime.TotalSeconds:0} seconds to {pSession} playtime");
+
+        foreach (var role in roles)
+        {
+            _roleTimers.AddTimeToRole(pSession.UserId, role, addedTime, dbSave);
+        }
+
+        _roleTimers.AddTimeToOverallPlaytime(pSession.UserId, addedTime, dbSave);
+        _lastSetTime[pSession] = currentTime;
+    }
+
+    private void SaveRoles(IPlayerSession pSession, TimeSpan currentTime, IEnumerable<string> roles, bool dbSave = true)
+    {
+        if (!_lastSetTime.TryGetValue(pSession, out var lastSave))
+        {
+            lastSave = currentTime;
+        }
+
+        if (currentTime <= lastSave) return;
+
+        var addedTime = currentTime - lastSave;
+        Sawmill.Info($"Adding {addedTime.TotalSeconds:0} seconds to {pSession} playtime");
+
+        foreach (var role in roles)
+        {
+            _roleTimers.AddTimeToRole(pSession.UserId, role, addedTime, dbSave);
+        }
+
+        _roleTimers.AddTimeToOverallPlaytime(pSession.UserId, addedTime, dbSave);
+        _lastSetTime[pSession] = currentTime;
+    }
```

Comment:
Look it's 1:30 AM here but these functions are *way* too similar so something is wrong here.